Added displayProgress method on MultipartUploader

cintiashamsu opened the pull request.

SamRemis commented on src/S3/ObjectUploader.php:
```diff
@@ -92,12 +92,16 @@ public function promise()
         }
-        return $this->client->executeAsync($command);
+//        return $this->client->executeAsync($command);
+        $test = $this->client->executeAsync($command);
```
Undo these changes

SamRemis commented on src/S3/MultipartUploader.php:
```diff
@@ -134,10 +152,22 @@ protected function createPart($seekable, $number)
+    {
+        $threshold = $this->progressThresholds;
+
+        if (!empty($threshold) and !empty($this->progressBar) and $this->uploadedBytes >= $threshold[0]) {
```
'and' is a lot more common in python; we use && in this SDK.  This should also be split into three lines

SamRemis replied:
This also needs to be a while loop; what happens when we pass two thresholds at once (IE 4 part uploads)

SamRemis commented on src/S3/ObjectUploader.php:
```diff
@@ -92,12 +92,16 @@ public function promise()
     public function upload()
     {
-        return $this->promise()->wait();
+        $result = $this->promise()->wait();
```
What is this change for?

stobrien89 commented on src/Multipart/AbstractUploader.php:
```diff
@@ -147,4 +147,7 @@ protected function getNumberOfParts($partSize)
         }
         return null;
     }
+
```
Remove newline here

stobrien89 commented on src/Multipart/AbstractUploader.php:
```diff
@@ -147,4 +147,7 @@ protected function getNumberOfParts($partSize)
+
 }
+
+
```
Also one of these

stobrien89 commented on src/Multipart/UploadState.php:
```diff
@@ -118,6 +118,7 @@ public function getUploadedParts()
      * @param int $status Status is an integer code defined by the constants
      *                    CREATED, INITIATED, and COMPLETED on this class.
      */
+
```
also this

stobrien89 commented on src/S3/MultipartUploadingTrait.php:
```diff
@@ -49,12 +51,24 @@ public static function getStateFromService(
 
+    protected function createProgressThresholds($totalSize)
+    {
+        $this->progressThresholds = [];
```
Is this defined anywhere in the classes this is used in?

stobrien89 commented on src/S3/MultipartUploadingTrait.php:
```diff
@@ -49,12 +51,24 @@ public static function getStateFromService(
+        for ($i=1;$i<=8;$i++) {
+            $this->progressThresholds []= round($totalSize*($i/8));
+        }
+        $this->getState()->setProgressThresholds($this->progressThresholds);
```
I think it would be better to set this in the constructor of `UploadState`.  That includes moving this logic to that class (`setProgressThresholds`).  Generally speaking, you want to set the initial values of your class members in the constructor of a given class.

stobrien89 reviewed (changes requested):
Just some whitespace changes again, but looks good otherwise!

stobrien89 commented on tests/Multipart/UploadStateTest.php:
```diff
@@ -70,4 +70,60 @@ public function testSerializationWorks()
+        return [
+            [100000, 0, $progressBar[0]],
+            [100000, 12499, $progressBar[0]],
+            [100000, 12500, "$progressBar[0]$progressBar[1]"],
```
in-string variables should be enclosed in braces for readability

stobrien89 replied:
could you add some more isolated/incremental cases to ensure we're hitting each threshold as expected?

stobrien89 commented on tests/Multipart/UploadStateTest.php:
```diff
@@ -70,4 +70,60 @@ public function testSerializationWorks()
+    /**
+     * @dataProvider getDisplayProgressCases
+     */
+    public function testDisplayProgressPrints(
```
could you rename this to `testDisplayProgressPrintsProgress`? we want test names to be as descriptive as possible

SamRemis commented on src/Multipart/UploadState.php:
```diff
@@ -13,6 +13,18 @@ class UploadState
     const INITIATED = 1;
     const COMPLETED = 2;
 
+    protected $progressBar = [
```
$progressBar and $progressThresholds should be combined into one array, with the keys being the thresholds and the values being what should be printed out.  One way of doing this would be array_combine.  Example:
[
0 => "Transfer initiated...\n|                    | 0.0%\n",
125 => "... 12.5%..."
//etc...
]

SamRemis commented on src/Multipart/UploadState.php:
```diff
@@ -25,12 +37,15 @@ class UploadState
     public function __construct(array $id)
     {
         $this->id = $id;
+        echo array_shift($this->progressBar);
```
This should be moved to be a part of the display progress method with a threshold of 0

stobrien89 reviewed (changes requested):
Left a couple of comments about the config options.  Let's chat about how this config will get plumbed through to `UploadState`

stobrien89 committed:
Merge branch 'master' into new-feature

aws-sdk-php-automation committed:
3.322.2 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.3 release

stchr committed:
fix typo in CHANGELOG.md (#2980)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.4 release

stobrien89 committed:
bugfix: LocationConstraint middleware (#3002)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.5 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.6 release

yenfryherrerafeliz committed:
chore: remove worklink client (#3005)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.7 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.8 release

Jubeki committed:
fix: PHP 8.4 deprecations with implicit nullable parameters (#2995)

Co-authored-by: Sean O'Brien <obrien.sean.dev@gmail.com>

aws-sdk-php-automation committed:
3.322.9 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.322.10 release

yenfryherrerafeliz committed:
enhancement: remove smoke test for autoscaling (#3007)

Remove smoke test for CreateLaunchConfiguration operation since it is not supported.
Please see: https://docs.aws.amazon.com/autoscaling/ec2/userguide/migrate-to-launch-templates.html

aws-sdk-php-automation committed:
3.323.0 release

aws-sdk-php-automation committed:
Update models for release

yenfryherrerafeliz committed:
enhacement: remove cloudsearch smoke test (#3009)

aws-sdk-php-automation committed:
3.323.1 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.323.2 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
chore: remove references to TranscribeStreaming (#3010)

RanVaknin committed:
adding return annotations to supress warnings (#3011)

aws-sdk-php-automation committed:
3.323.3 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.323.4 release

yenfryherrerafeliz committed:
chore: skip SubscriptionRequired exceptions (#3012)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.0 release

stobrien89 committed:
chore: add origin param to feedback url (#3013)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.1 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.2 release

stobrien89 committed:
chore: update `RemoveUnusedServices` README

aws-sdk-php-automation committed:
3.324.3 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.4 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.5 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.6 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.7 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.8 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.9 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.10 release

stobrien89 committed:
chore: remove deprecated nimble studio service (#3015)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.11 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.12 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.324.13 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.0 release

stobrien89 committed:
chore: remove call to cdn (#3019)

aws-sdk-php-automation committed:
3.325.1 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.2 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.3 release

stobrien89 committed:
chore: bundle fontawesome (#3021)

Co-authored-by: Sean O'Brien <obrien.sean.dev@gmail.com>

stobrien89 committed:
chore: remove fa cdn (#3022)

Co-authored-by: Sean O'Brien <obrien.sean.dev@gmail.com>

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.4 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.5 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.6 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.325.7 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.326.0 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.327.0 release

aws-sdk-php-automation committed:
3.327.1 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
chore: remove elastic transcoder smoke tests (#3028)

Co-authored-by: Sean O'Brien <obrien.sean.dev@gmail.com>

aws-sdk-php-automation committed:
3.328.0 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.328.1 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
chore: add query mode header (#3025)

aws-sdk-php-automation committed:
3.328.2 release

aws-sdk-php-automation committed:
Update models for release

rmn-eloha-jouny committed:
enhancement: adds support for psr/simple-cache v2 and v3 (#2838)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.328.3 release

stobrien89 committed:
bugfix: rest path resolution (#3032)

Co-authored-by: Sean O'Brien <obrien.sean.dev@gmail.com>

aws-sdk-php-automation committed:
3.329.0 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.330.0 release

stobrien89 committed:
bugfix: s3 path handling (#3038)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.330.1 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.330.2 release

issakujitsuk committed:
Add support for psr/cache 2.0, 3.0 (#3039)

yenfryherrerafeliz committed:
fix: honor use_aws_shared_config_files (#3027)

aws-sdk-php-automation committed:
3.331.0 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.332.0 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.332.1 release

aws-sdk-php-automation committed:
3.332.2 release

aws-sdk-php-automation committed:
Update models for release

beqqrry-aws committed:
Adds analytics to the base layout (#3041)

aws-sdk-php-automation committed:
3.333.0 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
bugfix: normalize rest base paths (#3042)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.334.0 release

aws-sdk-php-automation committed:
3.334.1 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.334.2 release

aws-sdk-php-automation committed:
3.334.3 release

aws-sdk-php-automation committed:
Update models for release

yenfryherrerafeliz committed:
chore: update changelong with breaking change (#3045)

Update changelong for `3.324.6 - 2024-10-18` with a more detailed breaking change message for the Athena service model change.

yenfryherrerafeliz committed:
enhancement: user agent 2.1 (#3001)

* enhancement: user agent 2.1

This change provides:
- A builder class for appending metrics
- Default initialization of a metrics builder within command instantiation
- Wraps user agent logic into a single middleware class
- Adds middlewares into the different features from where metrics can be gather, to acomplish this purpose.

* feat: user agent v2.1 second revision

- Move the user agent middleware after the signing step in order to gather signature metrics.
- Add request compression metric gathering.
- Add specific testing for signatures.
- Add specific testing for request compression.
- Add specific testing for s3 encryption clients.
- Add credentials metric gathering logic.
- Add tests for credentials metrics.
- Make existent credentials tests to work with the new field `source`. For example, for tests around profile credentials the source property for credentials MUST be `profile`. The default value for this field is `static`.

* chore: address PR feedback

- Use static mapping instead of multiple if/else-ifs.
- Remove the endpoint_id metric tracking since it was removed from the spec.
- Make method's name to be verb based.
- Remove the getMetricsBuilder from the CommandInterface.

* chore: add credentials->source tests

- Add tests cases to cover credentials sets the correct source when resolved.
- Fix the sso, sso_legacy, and process metrics. They should be all a profile based metric.

* chore: add missing blank lines

yenfryherrerafeliz committed:
Revert "enhancement: user agent 2.1 (#3001)" (#3046)

This reverts commit c9b20ac76063a437302c16b0a0dedd91b927e874.

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.334.4 release

aws-sdk-php-automation committed:
3.334.5 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.334.6 release

aws-sdk-php-automation committed:
3.334.7 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.335.0 release

aws-sdk-php-automation committed:
Update models for release

yenfryherrerafeliz committed:
enhancement: user agent 2.1 (#3047)

This change provides:
- A builder class for appending metrics
- Default initialization of a metrics builder within command instantiation
- Wraps user agent logic into a single middleware class
- Adds middlewares into the different features from where metrics can be gather, to acomplish this purpose.
- Move the user agent middleware after the signing step in order to gather signature metrics.
- Add request compression metric gathering.
- Add specific testing for signatures.
- Add specific testing for request compression.
- Add specific testing for s3 encryption clients.
- Add credentials metric gathering logic.
- Add tests for credentials metrics.
- Make existent credentials tests to work with the new field `source`. For example, for tests around profile credentials the source property for credentials MUST be `profile`. The default value for this field is `static`.
- Use static mapping instead of multiple if/else-ifs.
- Remove the endpoint_id metric tracking since it was removed from the spec.
- Make method's name to be verb based.
- Remove the getMetricsBuilder from the CommandInterface.
- Add tests cases to cover credentials sets the correct source when resolved.
- Fix the sso, sso_legacy, and process metrics. They should be all a profile based metric.

yenfryherrerafeliz committed:
chore: add changelogs (#3049)

Add changelogs for user agent 2.1 implementation.

yenfryherrerafeliz committed:
fix: add endpoint override flag (#3050)

* fix: add endpoint override flag

This change is needed because the parameter `endpoint` will always be populated even if a value was not provided at client construction, and
therefore we can not relay on it to determine its source.

* chore: fix endpoint override test

aws-sdk-php-automation committed:
3.336.0 release

aws-sdk-php-automation committed:
Update models for release

yenfryherrerafeliz committed:
fix: ddb account id metric test (#3051)

Basically, some e2e tests do not use the latest SDK version and therefore latest changes in the endpoint rule set file are not present there, which makes some of tests depending on those latest changes to fail. In this specific case it failed when checking if an account id was resolved or not in dynamodb, because the endpoint ruleset file did not contain the AccountId parameter. For fixing this I use a custom endpoint provider where the ruleset file contains the needed parameter and rule.

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.1 release

aws-sdk-php-automation committed:
3.336.2 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.3 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.4 release

aws-sdk-php-automation committed:
3.336.5 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.6 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.7 release

aws-sdk-php-automation committed:
3.336.8 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.9 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.10 release

aws-sdk-php-automation committed:
3.336.11 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.12 release

stobrien89 committed:
chore: remove iot 1click services (#3057)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.13 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.14 release

yenfryherrerafeliz committed:
fix: account_id resolution in ecs (#3058)

Co-authored-by: Sean O'Brien <60306702+stobrien89@users.noreply.github.com>

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.336.15 release

stobrien89 committed:
feat: flexible checksums v2 (#2999)

aws-sdk-php-automation committed:
3.337.0 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.337.1 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.337.2 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.337.3 release

stobrien89 committed:
feat: 8.0.x deprecation (#3061)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.338.0 release

aws-sdk-php-automation committed:
3.338.1 release

aws-sdk-php-automation committed:
Update models for release

yenfryherrerafeliz committed:
chore: account id endpoint metric refactor(#3055)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.338.2 release

stobrien89 committed:
feat: DSQL auth token generator (#3064)

aws-sdk-php-automation committed:
3.339.0 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
chore: update auth token generator default expiry time (#3067)

stobrien89 committed:
chore: update checksum-validation-config' (#3069)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.1 release

aws-sdk-php-automation committed:
3.339.2 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.3 release

stobrien89 committed:
Update README.md for 8.1+

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.4 release

aws-sdk-php-automation committed:
3.339.5 release

aws-sdk-php-automation committed:
Update models for release

dependabot[bot] committed:
Bump aws-actions/closed-issue-message from 1 to 2 (#3074)

stobrien89 committed:
bugfix: sso provider expiration (#3075)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.6 release

aws-sdk-php-automation committed:
3.339.7 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
chore: update illegal v4a headers (#3077)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.8 release

ziegenberg committed:
feat: remove paragonie/random_compat dev dependency (#3071)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.9 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.10 release

yenfryherrerafeliz committed:
feat: track checksum calculation validation mode (#3065)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.11 release

stobrien89 committed:
bugfix: auth fallbacks (#3082)

ziegenberg committed:
chore: run GHA HandleStaleDiscussions only on aws repo (#3072)

aws-sdk-php-automation committed:
3.339.12 release

aws-sdk-php-automation committed:
Update models for release

stobrien89 committed:
chore: endpoint arn middleware updates (#3084)

aws-sdk-php-automation committed:
3.339.13 release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.14 release

yenfryherrerafeliz committed:
fix: prevent duplicating metrics middleware (#3087)

aws-sdk-php-automation committed:
Update models for release

aws-sdk-php-automation committed:
3.339.15 release

stobrien89 committed:
Merge branch 'master' into new-feature
